doc/developers-guide/accumulators.md

```diff
@@ -30,132 +30,111 @@ a) performance - if we need to do something involving inspecting a stanza or mor
-Here we initialise the accumulator and pass it on.
+The rationale behind stripping an accumulator is that some values stored in it are context-dependend.
+For example, at the beginning `lserver` refers to the host of the sender C2S.
+When an accumulator goes to the c2s of the recipient, the `lserver` attribute may change.
+There are also many cached values which are not valid anymore when user changes (e.g. privacy checks).
 
-### `ejabberd_c2s:handle_incoming_message/3`
+In order to strip an accumulator, please use `strip(strip_params(), t())`, where `strip_params()` is a map of:
 
-This is called by `ejabberd_c2s:handle_info/3` when an accumulator from another c2s process is received.
-It is a ready-made accumulator, but stripped of all cached values; it preserves the original `ref`, `timestamp`, `from` and `to`, but doesn't have the `user` and `server` which have to be set here since they are likely to differ.
+* `lserver` - New host context. Obviously, may be equal to the old value.
+* `element`, `from_jid`, `to_jid` - The same rules apply as in `update_stanza/2`.
 
-### `ejabberd_s2s_in:route_incoming_stanza/4`
+## Main principles of an accumulator processing
 
-Here is where a stanza from another MongooseIM server is received.
+1. An accumulator is created when a stanza enters the server.
+2. XML stanza is never passed around as a pure `exml:element()`.
+3. An accumulator is stripped when it is passed to a different context (e.g. another c2s process).
+4. If a process produces more stanzas to be routed, they must reuse original acc but with stanza replaced with `update_stanza/2`.
 
 ## Hooks
 
 Many of the MongooseIM functionalities are implemented in submodules which attach their handlers to hooks (this is covered in detail in ["Hooks and handlers"](Hooks-and-handlers.md).
 When it comes to the accumulators, the following rules apply:
 
-* when using `run_fold` there are only a few cases in which something else should be passed as the accumulator.
- This is the case when for example handlers are supposed to modify the state (the state should be the accumulator).
- Your handlers should stash their results in the acc, so that you can extract what you need and pass the modified acc on.
-* avoid passing superfluous arguments to handlers - e.g. a `Server` in hook args is redundant since it is already present in the accumulator.
-* do not use `run` - it is against the main design principles.
+* When using `run_fold`, an accumulator should be provided every time the hook is connected to stanza processing. A hook handler may modify an accumulator in every permitted way (i.e. shouldn't directly modify acc fields, bypassing `mongoose_acc` API) and should return the execution result in `hook:result` field. This is not enforced but should be followed by convention.
```

Comment:
With every use of`run_fold`, an accumulator should be provided the hook connected to stanza processing. A hook handler may modify an accumulator in every permitted way (i.e. shouldn't directly modify acc fields, bypassing `mongoose_acc` API) and should return the execution result in the `hook:result` field. This is not enforced but should be followed by convention.

?